neuralpp/symbolic/basic_expression.py

```diff
@@ -13,23 +13,41 @@ def __init__(self, python_callable: Callable):
         super().__init__(f"{python_callable} is ambiguous.")
 
 
-def infer_python_callable_type(python_callable: Callable) -> ExpressionType:
+def boolean_function_of_arity(arity: int) -> Callable:
+    return Callable[[bool for i in range(arity)], bool]
+
+
+def infer_python_callable_type(python_callable: Callable, argument_types: List[ExpressionType] = None) -> \
+        Callable:
     match python_callable:
         # boolean operation
         case operator.and_ | operator.or_ | operator.xor:
-            raise AmbiguousTypeError(python_callable)  # this is also ambiguous because the arity could be arbitrary
-            # return Callable[[bool, bool], bool]
+            if argument_types is None:
+                raise AmbiguousTypeError(python_callable)  # this is also ambiguous because the arity could be arbitrary
+            if not all([argument_type == bool for argument_type in argument_types]):
+                raise TypeError(f"Argument types to boolean function {python_callable} should all be booleans. "
+                                f"Got {argument_types}.")
+            return boolean_function_of_arity(len(argument_types))
```

Comment:
Looks like we can also return `Callable[argument_types, bool]` since `argument_types` is guaranteed to be a list of `bool`s